src/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -48,83 +48,74 @@ private abstract class DiagnosticGetter(
             protected virtual bool ShouldIncludeDiagnostic(DiagnosticData diagnostic) => true;
 
             protected abstract Task ProduceDiagnosticsAsync(
-                Project project, IReadOnlyList<DocumentId> documentIds, bool includeProjectNonLocalResult, Action<DiagnosticData> callback, CancellationToken cancellationToken);
+                Project project, IReadOnlyList<DocumentId> documentIds, bool includeProjectNonLocalResult, ArrayBuilder<DiagnosticData> builder, CancellationToken cancellationToken);
 
             public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(CancellationToken cancellationToken)
             {
-                if (ProjectId != null)
-                {
-                    var project = Solution.GetProject(ProjectId);
-                    if (project == null)
-                        return [];
-
-                    // return diagnostics specific to one project or document
-                    var includeProjectNonLocalResult = DocumentId == null;
-                    return await ProduceProjectDiagnosticsAsync(
-                        [project], project => _getDocuments(project, DocumentId), includeProjectNonLocalResult, cancellationToken).ConfigureAwait(false);
-                }
+                var project = Solution.GetProject(ProjectId);
+                if (project == null)
+                    return [];
 
-                return await ProduceSolutionDiagnosticsAsync(Solution, cancellationToken).ConfigureAwait(false);
+                // return diagnostics specific to one project or document
+                var includeProjectNonLocalResult = DocumentId == null;
+                return await ProduceProjectDiagnosticsAsync(
+                    project, _getDocuments(project, DocumentId), includeProjectNonLocalResult, cancellationToken).ConfigureAwait(false);
             }
 
-            protected Task<ImmutableArray<DiagnosticData>> ProduceSolutionDiagnosticsAsync(Solution solution, CancellationToken cancellationToken)
-                => ProduceProjectDiagnosticsAsync(solution.Projects, static project => project.DocumentIds, includeProjectNonLocalResult: true, cancellationToken);
-
             protected async Task<ImmutableArray<DiagnosticData>> ProduceProjectDiagnosticsAsync(
-                IEnumerable<Project> projects, Func<Project, IReadOnlyList<DocumentId>> getDocumentIds,
+                Project project, IReadOnlyList<DocumentId> documentIds,
                 bool includeProjectNonLocalResult, CancellationToken cancellationToken)
             {
                 // PERF: run projects in parallel rather than running CompilationWithAnalyzer with concurrency == true.
```

Comment:
> // PERF: run projects in parallel rather than running CompilationWithAnalyzer with concurrency == true. [](http://example.com/codeflow?start=15&length=104)

seems like this comment is outdated now

Reply:
removed.